Adds helm chart support

DerekHeldtWerle opened the pull request:
*Issue #, if available:
This resolves https://github.com/opendistro-for-elasticsearch/community/issues/13

*Description of changes:*

This PR adds a configurable helm chart for opendistro elasticsearch that bundles the kibana and elasticsearch installation. The chart is based on the manifests provided [here](https://github.com/opendistro-for-elasticsearch/community/pull/45) and more directly incorporates features such as RBAC, TLS, and a number of other configurable options.

The way this chart is setup allows users to create a handful of secrets prior to issuing a `helm install` command, and in return will have a completely secure cluster with TLS for kibana, the elasticsearch transport/rest layers, the admin user when working with the `securityadmin.sh` script, as well the option to provide a custom version of the [securityconfig](https://github.com/opendistro-for-elasticsearch/security/tree/master/securityconfig) files without having to `exec` into any of the pods on startup. 

For the security config files, users can still make configuration changes as they see fit by following the instructions [here](https://github.com/opendistro-for-elasticsearch/community/tree/master/open-distro-elasticsearch-kubernetes#cluster-security-initialization) if that is the preferred route.


Note to reviewers:
I'm not 100% sure on exactly what is required on the license side of things in relation to all included files. Directing me on what needs to be done on that front would be greatly appreciated.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

alolita commented:
Hi @DerekHeldtWerle This is awesome! Thanks for contributing this PR. Will take a look and get back to you with questions. Let's work together on adding license, documentation etc. :-)

ChiefAlexander commented on open-distro-elasticsearch-kubernetes/helm/opendistro-es/templates/elasticsearch/master-svc.yaml:
```diff
@@ -0,0 +1,15 @@
+  name: {{ template "opendistro-es.fullname" . }}-discovery
+  namespace: {{ .Release.Namespace }}
+spec:
+  ports:
```
expose the metrics endpoint here

ChiefAlexander commented on open-distro-elasticsearch-kubernetes/helm/opendistro-es/templates/elasticsearch/es-data-svc.yaml:
```diff
@@ -0,0 +1,23 @@
+  ports:
+  - port: 9300
+    name: transport
+  - port: 9200
```
this isn't exposed and be be removed

ChiefAlexander commented on open-distro-elasticsearch-kubernetes/helm/opendistro-es/templates/elasticsearch/role.yaml:
```diff
@@ -0,0 +1,15 @@
+{{ include "opendistro-es.labels.standard" . | indent 4 }}
+rules:
+- apiGroups: ['extensions']
+  resources: ['podsecuritypolicies']
```
Completely missed the psp file in the directory above 🤦‍♂

DerekHeldtWerle commented on open-distro-elasticsearch-kubernetes/helm/opendistro-es/templates/elasticsearch/es-data-sts.yaml:
```diff
@@ -0,0 +1,176 @@
+        image: {{ .Values.elasticsearch.image }}:{{ .Values.elasticsearch.imageTag }}
+        imagePullPolicy: Always
+        # only publish the transport port
+        ports:
```
Admittedly, I used the [manifests](https://github.com/opendistro-for-elasticsearch/community/tree/master/open-distro-elasticsearch-kubernetes) as the source of truth for this chart. It does, however, look like there are some inconsistencies between the exposed ports and the services as you have pointed out. I haven't personally used the metrics endpoint with Performance Analyzer. Is that something you expose to all of the services, or just the client?

dpotapov commented:
I've tried to use the chart, but didn't get it working :-( Deployed containers are crashing without a reason...

In general, I found it much harder to configure it than stable/elasticsearch chart. Had to modify "fixmount" container to ignore the ".snapshot" path [data-statefulset.yaml#L93](https://github.com/helm/charts/blob/2b23dae64b8b04ee985ea55e1763351f66b1fd2c/stable/elasticsearch/templates/data-statefulset.yaml#L93). Without `elasticsearch.config` defined, containers are throwing an errror something like "tee: /usr/share/elasticsearch/config/elasticsearch.yml: Read-Only filesystem".

DerekHeldtWerle commented:
Hi, @dpotapov to support any potential configuration setup, I went with an empty default `elasticsearch.config`, `kibana.config`, `elasticsearch.log4jConfig`, and `elasticsearch.loggingConfig`. With this, as you encountered, if no configuration is defined there will be issues as the contents provided are directly mounted to their corresponding files in the container. 

Would it be preferred to only overwrite any of the aforementioned configs if a value was provided in the `values.yaml`? This would leave the default configuration files in tact, as defined by the docker images, while leaving it up to the user define their own configuration that would overwrite some or all of the files.

I'm not as well versed on the `.snapshot` issue. Do you mind expanding on that?

dpotapov commented:
If the default OpenDistro's configuration works well in clustered container environment, maybe it makes sense just leave it as-is (unless specific config is provided in `values.yml`). Helm chart from the [stable repository](https://github.com/helm/charts/tree/master/stable/elasticsearch/) uses own pre-defined configuration in ConfigMap., while [Elastic's helm chart](https://github.com/elastic/helm-charts/tree/master/elasticsearch) doesn't. Though it allows to override files under elasticsearch's config directory with ConfigMap as well.

Regarding the `.snapshot`: this is relevant if a NAS storage is used as a back-end for persistent volumes in Kubernetes. Snapshots is a feature of some NAS devices (e.g. NetApp, EMC) to provide access to files on a NFS volume at some point of time in the past. Those files are accessible read-only (so all chown operations will fail).

DerekHeldtWerle commented:
The default configuration files baked into the docker images in question:

```
kibana.yml:

## Kibana
# Default Kibana configuration from kibana-docker.

server.name: kibana
server.host: "0"
elasticsearch.url: https://localhost:9200
elasticsearch.ssl.verificationMode: none
elasticsearch.username: kibanaserver
elasticsearch.password: kibanaserver
elasticsearch.requestHeadersWhitelist: ["securitytenant","Authorization"]

opendistro_security.multitenancy.enabled: true
opendistro_security.multitenancy.tenants.preferred: ["Private", "Global"]
opendistro_security.readonly_mode.roles: ["kibana_read_only"]
```
```
elasticsearch.yml:

cluster.name: "docker-cluster"
network.host: 0.0.0.0

# minimum_master_nodes need to be explicitly set when bound on a public IP
# set to 1 to allow single node clusters
# Details: https://github.com/elastic/elasticsearch/pull/17288
discovery.zen.minimum_master_nodes: 1
```
There's no default `elasticsearch.logging` config.

```
log4j2.properties
status = error

appender.console.type = Console
appender.console.name = console
appender.console.layout.type = PatternLayout
appender.console.layout.pattern = [%d{ISO8601}][%-5p][%-25c{1.}] [%node_name]%marker %m%n

rootLogger.level = info
rootLogger.appenderRef.console.ref = console
```

A hybrid approach might then be a good option where the `elasticsearch.yml` and `logging.yml` by default can more so mirror the configuration from the [manifests](https://github.com/opendistro-for-elasticsearch/community/blob/master/open-distro-elasticsearch-kubernetes/elasticsearch/30-es-configmap.yml#L50) (TLS options removed) with the ability to overwrite it if desired. The `kibana.yml` seems ok (correct me if I'm wrong here) in terms of a default insecure installation and the `log4j2.properites` seems fine as well.

Proposed updates to `values.yaml`:
```
elasticsearch:
  # Default configuration
  config:
    cluster:
      name: ${CLUSTER_NAME}
    node:
      master: ${NODE_MASTER}
      data: ${NODE_DATA}
      name: ${NODE_NAME}
      ingest: ${NODE_INGEST}
      max_local_storage_nodes: 1
      attr.box_type: hot
    processors: ${PROCESSORS:1}
    network.host: ${NETWORK_HOST}
    thread_pool.bulk.queue_size: 800
    path:
      data: /usr/share/elasticsearch/data
      logs: /usr/share/elasticsearch/logs
    http:
      enabled: ${HTTP_ENABLE}
      compression: true
    discovery:
      zen:
        ping.unicast.hosts: ${DISCOVERY_SERVICE}
        minimum_master_nodes: ${NUMBER_OF_MASTERS}

    opendistro_security.audit.type: internal_elasticsearch
    opendistro_security.enable_snapshot_restore_privilege: true
    opendistro_security.check_snapshot_restore_write_privileges: true
    opendistro_security.restapi.roles_enabled: ["all_access", "security_rest_api_access"]
    cluster.routing.allocation.disk.threshold_enabled: false
    opendistro_security.audit.config.disabled_rest_categories: NONE
    opendistro_security.audit.config.disabled_transport_categories: NONE

  loggingConfig:
   # Default config
    # you can override this using by setting a system property, for example -Des.logger.level=DEBUG
    es.logger.level: INFO
    rootLogger: ${es.logger.level}, console
    logger:
      # log action execution errors for easier debugging
      action: DEBUG
      # reduce the logging for aws, too much is logged under the default INFO
      com.amazonaws: WARN
    appender:
      console:
        type: console
        layout:
          type: consolePattern
          conversionPattern: "[%d{ISO8601}][%-5p][%-25c] %m%n"
```

This _should_ allow for a minimal installation to go through, and give the user the option to further configure if desired.

DerekHeldtWerle commented:
After doing some digging, it looks like there are some demo certs (and additional configuration) in the docker image that are installed at startup. The entrypoint for the docker container includes:
```
if [[ -d "/usr/share/elasticsearch/plugins/opendistro_security" ]]; then
    # Install Demo certifactes for Security Plugin and update the elasticsearch.yml
    # file to use those certificates.
    /usr/share/elasticsearch/plugins/opendistro_security/tools/install_demo_configuration.sh -y -i -s
fi
```
With the above arguments passed to the [install_demo_configuration](https://github.com/opendistro-for-elasticsearch/security/blob/dfc41db0d0123cd0965d40ee47d61266e560f7e6/tools/install_demo_configuration.sh), these certs should then already be accessible for use. The script then also configures the elasticsearch.yml to a greater extent than what I had previously linked above. By running the above script with the provided arguments, you will receive an `elasticsearch.yml` that looks like the following:

```
cluster.name: "docker-cluster"
network.host: 0.0.0.0

# minimum_master_nodes need to be explicitly set when bound on a public IP
# set to 1 to allow single node clusters
# Details: https://github.com/elastic/elasticsearch/pull/17288
discovery.zen.minimum_master_nodes: 1

######## Start OpenDistro for Elasticsearch Security Demo Configuration ########
# WARNING: revise all the lines below before you go into production
opendistro_security.ssl.transport.pemcert_filepath: esnode.pem
opendistro_security.ssl.transport.pemkey_filepath: esnode-key.pem
opendistro_security.ssl.transport.pemtrustedcas_filepath: root-ca.pem
opendistro_security.ssl.transport.enforce_hostname_verification: false
opendistro_security.ssl.http.enabled: true
opendistro_security.ssl.http.pemcert_filepath: esnode.pem
opendistro_security.ssl.http.pemkey_filepath: esnode-key.pem
opendistro_security.ssl.http.pemtrustedcas_filepath: root-ca.pem
opendistro_security.allow_unsafe_democertificates: true
opendistro_security.allow_default_init_securityindex: true
opendistro_security.authcz.admin_dn:
  - CN=kirk,OU=client,O=client,L=test, C=de

opendistro_security.audit.type: internal_elasticsearch
opendistro_security.enable_snapshot_restore_privilege: true
opendistro_security.check_snapshot_restore_write_privileges: true
opendistro_security.restapi.roles_enabled: ["all_access", "security_rest_api_access"]
cluster.routing.allocation.disk.threshold_enabled: false
node.max_local_storage_nodes: 3
######## End OpenDistro for Elasticsearch Security Demo Configuration ########
```

With this newfound information, and the [Dockerfiles not being available](https://github.com/opendistro-for-elasticsearch/community/issues/11) I now realize that mounting in the config is a requirement to prevent this startup script from overwriting any desired `elasticsearch.yml` options provided by the helm chart. If you use the above configuration to `elasticsearch.config` are your SSL issues will be resolved? If so, by default, we could then leave the `elasticsearch.yml` alone similar to the `kibana.yml` and the `log4j2.properties`.

DerekHeldtWerle commented:
@dpotapov I've updated the chart and was able to test out a default installation with no custom configurations provided to any of the options we have discussed. 

The `kibana.yml`, `elasticsearch.yml`, and `log4j2.properties` are now only conditionally mounted in allowing the `install_demo_configuration` script to successfully create and install the demo certs, and leaves the default configurations for each intact unless specified by the user. 

The `logging.yaml` config is now in the `values.yaml` by default to allow for custom configuration as it 
does exist in the docker container.

dpotapov commented:
I confirm that 921f13ed569d51de6d0571b5848f805e21d48799 works with SSL out-of-the-box. Though cluster reports "yellow" state:
```
# curl -k -i -u admin:admin https://127.0.0.1:9200/_cluster/health
HTTP/1.1 200 OK
content-type: application/json; charset=UTF-8
content-length: 403

{"cluster_name":"docker-cluster","status":"yellow","timed_out":false,"number_of_nodes":1,"number_of_data_nodes":1,"active_primary_shards":6,"active_shards":6,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":5,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":54.54545454545454}
```

kalvinnchau commented:
@dpotapov I pushed a commit that sets the `discovery.zen.ping.unicast.hosts` environment variable so that each node in the cluster is able to look for each other. The start up script looks for `.` separated variables and passes them via `-E` option at start-up.

We saw the same issue in our tests and found that each node was basically acting as its own cluster, and were not discovering each other (noted by the `"number_of_nodes":1` in your output. That would cause the indices that need to distribute its shards to never be able to do so. Now that the cluster nodes are aware of each other you should be able to have a green cluster.

This also fixes the resource `request` and `limits` being properly set.

dpotapov commented:
Can't get it working:
```
#  curl -k -u admin:admin https://127.0.0.1:9200/_cluster/health
{"cluster_name":"docker-cluster","status":"green","timed_out":false,"number_of_nodes":1,"number_of_data_nodes":1,"active_primary_shards":1,"active_shards":1,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":0,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":100.0}
```

It reports as "green", but `number_of_nodes` is still 1. Client nodes are constantly rebooting and not printing log messages.

kalvinnchau commented:
@dpotapov Sorry about that, the `Xmx` heap settings were at `512m` but the pod resource was only grabbing `128mi`. So your pods were most likely OOM'ing.

The latest commit should fix that and set more correct settings.

dpotapov commented:
Thanks, @kalvinnchau. After replacing "1024mi" with "1024Mi" it worked.

```
 v1.ResourceRequirements.Requests: Limits: unmarshalerDecoder: unable to parse quantity's suffix, error found in #10 byte of ...|":"1024mi"},"request|..., bigger context ...|],"resources":{"limits":{"cpu":1,"memory":"1024mi"},"requests":{"cpu":"200m","memory":"1024Mi"}},"vo|...
```

Is it ok when it says that there are 8 data nodes? should be just 3, isn't it?

```
{
"cluster_name": "docker-cluster",
"status": "green",
"timed_out": false,
"number_of_nodes": 8,
"number_of_data_nodes": 8,
"active_primary_shards": 1,
"active_shards": 8,
"relocating_shards": 0,
"initializing_shards": 0,
"unassigned_shards": 0,
"delayed_unassigned_shards": 0,
"number_of_pending_tasks": 0,
"number_of_in_flight_fetch": 0,
"task_max_waiting_in_queue_millis": 0,
"active_shards_percent_as_number": 100
}
```

kalvinnchau commented:
@dpotapov  Fixed that typo :)

The default config doesn't actually set whether or not the node is a `master`, `client` or `data` node. So by default, each node thinks its also a `data` node.

We could set that setting by default for each of the node types, but we wanted it to be as close to the generated defaults as possible.

For testing purposes it doesn't really hurt, any each node has the exact same defaults for resources so it wont make a difference. Production use cases will want to separate them out however.

dpotapov commented:
Hi,
With latest changes I've got Elasticsearch working, now dealing with Kibana.

* Installing a new Helm release fails with error:
  ```
  Error: release db3 failed: secrets "elasticsearch-config" already exists
  ```
* DevTools page in Kibana doesn't work: when running any query it fails with time-out:
  ```
  {
    "statusCode": 504,
    "error": "Gateway Time-out",
    "message": "Client request timeout"
  }
  ```
* After testing the chart, I've completely broke the Helm/Tiller. Any `helm install` command fails with:
  ```
  NAME:   db4
  Error: getting deployed release "db4": release: "db4" not found
  ```

dpotapov commented:
Upd: the issue about "error: release not found" was caused by too big ConfigMap, here is a log entry from tiller:

```
create: failed to create: ConfigMap "db4.v1" is invalid: []: Too long: must have at most 1048576 characters
```

I wonder whether it is better to split the chart into 2 separate ones: one for Elasticsearch and the second for Kibana.

dpotapov commented:
One more issue: upgrading the chart (via `helm upgrade`) with a new config file puts the cluster into a red state.

dpotapov commented:
To me it makes more sense to adopt https://github.com/elastic/helm-charts for OpenDistro, because their chart is much more mature (though they claim it is in "alpha" state).
I thought it would be difficult to use elastic's chart because of xpack vs. opendistro_security differences, but in fact it was not:

*values-elastic.yml*
```
volumeClaimTemplate:
  storageClassName: nfs
maxUnavailable: 0
image: amazon/opendistro-for-elasticsearch
imageTag: 0.9.0
protocol: https
resources:
  limits:
    cpu: 1
    memory: 2Gi
extraEnvs:
  - name: ELASTIC_USERNAME
    value: admin
  - name: ELASTIC_PASSWORD
    value: admin
```

*values-kibana.yml*
```
image: "amazon/opendistro-for-elasticsearch-kibana"
imageTag: "0.9.0"
elasticsearchURL: "https://elasticsearch-master:9200"
extraEnvs:
  - name: ELASTICSEARCH_USERNAME
    value: kibanaserver
  - name: ELASTICSEARCH_PASSWORD
    value: kibanaserver
```

DerekHeldtWerle commented:
Hey @dpotapov, with the latest commit helm upgrades are now handled properly. 

Going through elastic's helm charts, it does appear they do a number of things to get the default installation configured, which is an area this chart could definitely improve upon. We work in an environment where rbac, pod security policies, tls, multitenancy, etc. are required, and haven't had the time to get this chart to work with a minimal installation. All of the aforementioned options are first class citizens in this chart and mirror the goal of this chart in creating a handful a secrets, enabling the various security options in the config, and in return have a production level cluster with authn/z, multitenancy, and encryption across the board. That isn't to say there is no plan to support a default installation, it just didn't fall in line with our initial needs for the chart. In fact, we would be very receptive towards getting the default installation in a stable state from outside help.

On a personal note, we have seen initial success with our current cluster with tls enabled across the board and the setting of the various securityconfig components (ldap for authn/z, roles, roles mappings, tenants) that get initialized at startup via secrets we have added to the namespace, and would be happy to share our configs/add them to the readme if desired.

fgimenezm commented:
It's great to see this being developed!
Just two comments here:
 - why using secrets instead of configmaps for the config files?
 - have you seen this other chart from search-guard? It does some interesting things to handle certificates: https://github.com/floragunncom/search-guard-helm/tree/master/sg-helm

DerekHeldtWerle commented:
@fgimenezm I don't think I have a real reason why I chose the config files being secrets vs. configmaps. We could definitely interchange the two and the chart should still work fine. 

Related to the certs with search guard, are you referring to how they handle default certs? [Further up](https://github.com/opendistro-for-elasticsearch/community/pull/56#issuecomment-490531534) in this PR we found out that default certs for opendistro get generated at startup if `opendistro_security` is not found in the `elasticsearch.yml` file via [this script](https://github.com/opendistro-for-elasticsearch/security/blob/dfc41db0d0123cd0965d40ee47d61266e560f7e6/tools/install_demo_configuration.sh). 

From what we've seen, these default certs seem to do the trick on the tls front. The latest issue for the default installation is the kibana dev tools timing out.

kfox1111 commented:
Do the config files contain any secrets?

Has there been any effort to integrate with cert-manager for managing the certificates? I've used it for a while now and its really nice.

DerekHeldtWerle commented:
Technically speaking, yes, a user could set the `elasticsearch.username` or `elasticsearch.password` directly in `kibana.yml`, as well as some [audit options](https://github.com/opendistro-for-elasticsearch/security/blob/master/securityconfig/elasticsearch.yml.example#L92) that would be considered secrets. Since the variables of form `${VAR}` get expanded though, best practice would likely be to pass it through the environment. With this, we can add in [extraEnvs](https://github.com/elastic/helm-charts/blob/master/elasticsearch/templates/statefulset.yaml#L199) to allow for any number of expandable options via the environment to this chart. Just to be extra safe, it might be best to keep these configs as secrets for users in multi-tenant environments.

On the certs front, we have an internal authority so I'm not sure what sort of effort would be needed on that front to get it to work with cert-manager.

hkroger commented:
I see that pods are missing release labels in ES master deployment:

```
Start Time:         Fri, 17 May 2019 10:24:14 +0000
Labels:             pod-template-hash=6c75b6f4dc
                    role=master
Annotations:        <none>
Status:             Running
```

I don't know if it is missing elsewhere.

hkroger commented:
It's a matter of taste but is it necessary to have the suffix "-svc" for services? I'm fine either way, just pointing it out. :)

hkroger commented:
Another thing: some more parameters might be useful to be brought to kibana values. E.g. server.basePath is useful when running kibana behind a proxy in a subpath.

DerekHeldtWerle commented:
Hey @hkroger nice finds! I've fixed the labeling issues and added in the option to specify `extraEnvs` for both elastic and kibana. 

Related to setting `server.basePath` where are you looking to set that? Currently, you could set that under `kibana.config` and now you are able to set the matching environment variable via `extraEnvs`, but I'm not quite sure if that is what you are looking for.

hkroger commented:
@DerekHeldtWerle that's good! extraEnvs solves my problem at least so I'm happy with that approach.

hkroger commented:
One more thing, I saw that images are being hard coded at the moment. It would be useful if those were configurable. The reason is that we don't always have access to docker hub in some environments and there we will use different image configurations. Same goes with the pullSecrets. In some environments those are needed.

This is an example how that is implemented: https://github.com/helm/charts/blob/master/stable/postgresql/templates/statefulset.yaml

DerekHeldtWerle commented:
@hkroger The latest commit should now properly support both `imagePullSecrets` and a configurable `initContainer` image. 

I also ended up mirroring the way the postgresql chart used `global` values as that looks to be a more clean approach from a user perspective.

hkroger commented:
@DerekHeldtWerle I was checking Kibana deployment and I see that Kibana is missing liveness / readiness probes. ReadinessProbe is not used in elasticsearch either.

hkroger commented on open-distro-elasticsearch-kubernetes/helm/opendistro-es/templates/elasticsearch/es-client-deploy.yaml:
```diff
@@ -0,0 +1,173 @@
+        livenessProbe:
+          tcpSocket:
+            port: transport
+          initialDelaySeconds: 60
```
periodSeconds would be nice to define as well

nixgadget commented:
@DerekHeldtWerle I like where this is heading. Few things to improve however,
- I'd like to see the option of generating the certs automatically. I suppose there's various ways to achieve this. One of which is implemented by the search-guard folks in https://github.com/floragunncom/search-guard-helm/blob/master/sg-helm/templates/sgadmin-deployment.yaml#L53. The other alternative is to implement in a `job` and managed via Helm hooks. I like the later as it is much more cleaner to manager than a `deployment`.
- I know the Opendistro docker image itself doesnt provide any snapshot plugins out of the box but would be nice to implement it so users have some flexibility in terms of backups/restores.

kfox1111 commented:
>  
>     * I'd like to see the option of generating the certs automatically. I suppose there's various ways to achieve this. One of which is implemented by the search-guard folks in https://github.com/floragunncom/search-guard-helm/blob/master/sg-helm/templates/sgadmin-deployment.yaml#L53. The other alternative is to implement in a `job` and manage via Helm hooks. I like the later as it is much more cleaner to manager than a `deployment`.

Cert-manager was built to give k8s a declarative api to do exactly that. I've had really good luck with it and feels far less hacky then jobs. It also supports stuff like automatic rolling.

DerekHeldtWerle commented:
Hey @kfox1111, I spent a little time looking over cert-manager, and from what I'm seeing, it appears you just need to add a few  annotations to the `ingress` resource and specify the matching `secret` that housed the cert created to get it to work? If so, this should already be supported, but let me know if I'm missing something on that front.

mpucholblasco commented on open-distro-elasticsearch-kubernetes/helm/opendistro-es/templates/elasticsearch/es-data-sts.yaml:
```diff
@@ -0,0 +1,189 @@
+    spec:
+{{- include "opendistro-es.imagePullSecrets" . | indent 6 }}
+      # Add toleration for not scheduling on dedicated node
+      tolerations:
```
It's better to allow users to define their own `tolerations` in order to execute ES in a dedicated set of nodes. It also requires `nodeSelector` or another way to define in which nodes should pods run (such as `affinity`).

mpucholblasco commented on open-distro-elasticsearch-kubernetes/helm/opendistro-es/templates/kibana/kibana-service.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
```
Include `annotations` in order to allow users to define extra annotations for external LB.

kfox1111 commented:
You can use cert manager to secure the ingress via annotations. but you also can use it to provision ca's, request individual certs that are valid via lets-encrypt, etc. cert-manager itself secures its own web hook by making a self signed ca using itself, creates a ca issuer, and secure the communication between k8s and the hook with it. You can see some of that here (https://github.com/jetstack/cert-manager/blob/master/deploy/charts/cert-manager/webhook/templates/pki.yaml)

DerekHeldtWerle commented:
@mpucholblasco I've added in custom `annotation` and service `type` support for kibana, `nodeSelector` and custom `tolerations` across the board.

@hkroger I've updated the chart to allow for custom `readinessProbe`'s and `livenessProbe`'s on each individual service.

Reach out if anyone has any issues with the above additions.

mpucholblasco commented:
Perfect @DerekHeldtWerle , thanks a lot!

justinrush commented:
Took a bit, but did get it deployed with custom signed certs.  Likely worth adding some documentation about how the certs all work in the values.yaml, even though its really an open distro issue.

The values makes it seem like you can enable ssl just for admin and rest and not for transport, but the opendistro_security.ssl.http.enabled flag turns on ssl for all of them.

The client cert must be a full cert chain - cert, intermediates, then root.  That's not really intuitive since there is a specific secret key for the root ca, which causes an error if you try and include multiple certs in there.

I needed the following values overrides - would be great if the helm chart could generate this config when you flip the ssl enabled flags rather than having the user dig through sparse documentation to find the right settings.
kibana.config.elasticsearch.ssl.certificateAuthorities: /usr/share/kibana/certs/elk-rest-crt.pem
kibana.ssl.elasticsearch.enabled: true
kibana.ssl.elasticsearch.existingCertSecret: opendistro-es-dev-certs
elasticsearch.ssl.transport.enabled: true
elasticsearch.ssl.transport.existingCertSecret: opendistro-es-dev-certs
elasticsearch.ssl.rest.enabled: true
elasticsearch.ssl.rest.existingCertSecret: opendistro-es-dev-certs
elasticsearch.ssl.admin.enabled: true
elasticsearch.ssl.admin.existingCertSecret: opendistro-es-dev-certs
\# if you don't have the below, the nodes have no way of validating the cert is good and throw "certificate_unknown" errors.
opendistro_security.nodes_dn:
      - "CN=opendistro-es-dev.k8s.test.lab, OU=...
opendistro_security.ssl.transport.pemcert_filepath: elk-transport-crt.pem
opendistro_security.ssl.transport.pemkey_filepath: elk-transport-key.pem
opendistro_security.ssl.transport.pemtrustedcas_filepath: elk-transport-root-ca.pem
opendistro_security.ssl.transport.enforce_hostname_verification: false
opendistro_security.ssl.http.enabled: true
opendistro_security.ssl.http.pemcert_filepath: elk-rest-crt.pem
opendistro_security.ssl.http.pemkey_filepath: elk-rest-key.pem
opendistro_security.ssl.http.pemtrustedcas_filepath: elk-rest-root-ca.pem


Also - the containers complain about insecure cert permissions, they should be 0600, but they get owned by root.  I think you can modify the mount to specify the default mode and also the uid owner.

tfrench93 commented:
@justinrush Any chance you could share your values file? I am also trying to get this to work using self-signed certs and I am thoroughly stuck.

DerekHeldtWerle commented:
@tfrench93 I've updated the readme to include example secure configs for both kibana and elasticsearch, and added some info about the default installation as well. 

Let us know if these examples work for you to get your cluster up and running using self signed certs.

sc7565 commented:
@DerekHeldtWerle using default configurations, I am getting below error in Kibana 

`{"type":"log","@timestamp":"2019-06-14T16:31:06Z","tags":["warning","elasticsearch","admin"],"pid":1,"message":"Unable` to revive connection: https://elastic-platform-monitoring-client-service:9200/"}
{"type":"log","@timestamp":"2019-06-14T16:31:06Z","tags":["warning","elasticsearch","admin"],"pid":1,"message":"No living connections"}
{"type":"log","@timestamp":"2019-06-14T16:31:06Z","tags":["status","plugin:elasticsearch@6.7.1","error"],"pid":1,"state":"red","message":"Status changed from yellow to red - Unable to connect to Elasticsearch.","prevState":"yellow","prevMsg":"Waiting for Elasticsearch"}`

justinrush commented:
@DerekHeldtWerle @tfrench93 
Thanks, that did help.  The only other things that may be worth documenting is that the certs need to have both TLS Web Server Authentication and TLS Web Client Authentication x509v3 extended key attributes.

Also - the client workload needs to be modified to mount the certs with mode 660.  The helm chart mounts them with 600 and they are owned by root, but some of the processes running are running as the elasticsearch user which can't read the certs.

@tfrench93 let me know if you still have problems after reading the new readme commit

shamsalmon commented:
Anyone get this after configuring your certs? For reference I am using lets-encrypt for my rest certs (so no ca.crt) and self-signed for node certs. Everything works but my client container is going crazy.

```
[2019-06-18T03:41:40,783][ERROR][c.a.o.s.s.t.OpenDistroSecuritySSLNettyTransport] [pioneering-giraffe-opendistro-es-client-6bc85446b-gtcqn] SSL Problem Empty server certificate chain
javax.net.ssl.SSLHandshakeException: Empty server certificate chain
```

DerekHeldtWerle commented:
Hey @sc7565 and @dpotapov Kibana should now work fine with the default installation thanks to @dcuenot's PR.

I haven't had as much time to dedicate to this as of late, but will be testing out the cert permissions (@justinrush) and adding more docs on the requirements for the certs soon. 

As an aside, we are receptive towards any PR's made against our fork (thanks again @dcuenot!) and appreciate all of comments already made on this PR that have helped further its functionality and increase its robustness.

justinrush commented:
@DerekHeldtWerle: I'll try and generalize our values overrides for the group along with a large readme I put together, this weekend.

We've been doing some failure testing and have found that if we kill all the master nodes, when they come back up we have a cluster, but all our data appears to be gone... we have to re-create indexes, etc.  Found it by accident because we were deploying our dev cluster in a 3 node cluster, and with the anti-affinity rules, any upgrade to the chart would get stuck indefinitely, so during our testing period we'd just kill the pods and let them re-create.  We can certainly adjust the number of master nodes and change our upgrade strategy, but it is concerning that losing all the master nodes makes all the data go away - what is persistence for if it doesn't survive a reboot?  Is there a persistence option for master nodes that I'm missing that has shard data on it?

(sorry in advance for the naive elastic questions)

aegershman commented:
Any thoughts on when this might get merged in? It's <0.x.x, so the expectation is that the API could change from 0.0.1 to 0.0.2. Thoughts on getting things merged in and making iterations from there?

fsschmitt commented:
> As an aside, we are receptive towards any PR's made against our fork (thanks again @dcuenot!) and appreciate all of comments already made on this PR that have helped further its functionality and increase its robustness.

@DerekHeldtWerle FYI, I've created a PR on your fork (Viasat:feat/helm-chart) with some minor fixes regarding making the templates between client, data and master more standard and a fix related to a typo on a volume mount reference regarding action-groups.

DandyDeveloper commented:
I'm just spending some time getting up to speed on this. Will contribute as I get to grips with the state. 

What's the general view on how the certificates are currently managed through this chart? I see we've got the demo SSL stuff setup by default. 

If we want to test out getting scraped logs into ES (via Fluentd etc.) does anyone have a good idea on how this can be achieved with the default state of the helm chart?

hkroger commented:
Yes, I agree with all the SSL worries here since I spent myself about 2 weeks to get this working with Keycloak + custom certs all over the place. Full SSL examples with and without openid connector examples would be great here. They dont' need to be part of this PR but in general it would be very useful.

DandyDeveloper commented:
@justinrush Do you have an example or reference for how you generated the self-signed certificates?

Despite having a sign cert, with the correct root and correct CN in the `elasticsearch.yml` it's refusing to cooperate and throwing `SSL Problem Received fatal alert: certificate_unknown`.

DerekHeldtWerle commented:
Following @fsschmitt's PR, it became evident that matching the way [elastic's charts](https://github.com/elastic/helm-charts/blob/master/elasticsearch/templates/statefulset.yaml#L178) handle environment variables is the optimal route for the default installation due to the startup script [reading in](https://github.com/opendistro-for-elasticsearch/opendistro-build/blob/master/elasticsearch/docker/build/elasticsearch/bin/docker-entrypoint.sh#L51) environment variables of the form `x.y` and passing them in as options [on startup](https://github.com/opendistro-for-elasticsearch/opendistro-build/blob/master/elasticsearch/docker/build/elasticsearch/bin/docker-entrypoint.sh#L97). 

With these changes, the default installation now yields pods running the expected dedicated services (`client`, `master`, `data`). I also added in some docs on the requirement of keys to be PKC#5 and a link to amazon's cert setup docs for opendistro.

@DandyDeveloper I didn't personally run into those issues, but the [amazon blog](https://aws.amazon.com/blogs/opensource/add-ssl-certificates-open-distro-for-elasticsearch/) might be able to point you in the right direction.

DandyDeveloper commented:
@DerekHeldtWerle and others. Had you had any success with indexes being made? Mine are just solid red from a fluentd instance. 

`unexpected error during the primary phase for action [indices:data/write/bulk[s]], request [BulkShardRequest [[logstash-2019.06.27][2]] containing [2] requests]`

justinrush commented:
@DandyDeveloper When you create the cert, make sure it has both your FQDN and your short name in the SAN.  The certificate needs to have both "TLS Web Server Authenication" and "TLS Web Client Authentication" x509v3 extended key usage attributes.  You will end up with three secret key values per cert type (i just used the same cert for all three):

1. The cert (e.g. elk-transport-crt.pem).  This should have your cert up top, followed by any intermediates or subs, followed by your root.
2. The key (e.g. elk-transport-key.pem).  This should be the private key
3. The root ca (e.g. elk-transport-root-ca.pem).  This should be just the root CA (not subs)

@DerekHeldtWerle I think I finally tracked down my issues with regards to data resiliency.  I was having issues when I killed all the masters, the new ones would spin up, but all my data was gone. (Understand one shouldn't kill all the masters, but we were doing failure testing).  The demo config generates a config value for elastic called "node.max_local_storage_nodes" with a value of 3.  According to the [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html#max-local-storage-nodes) this allows multiple nodes to share the same datapath.  I don't pretend to understand why this causes my issues, but setting that to 1 allows me to kill all my master nodes and keep my data.

justinrush commented:
@DandyDeveloper There is another thing that was giving me the unknown cert error.  When you sent "opendistro_security.nodes_dn", you can't actually try and specify your node dn.  I tried all sorts of wildcards and settled on this as the only thing that works.

> opendistro_security.nodes_dn:
> \- "/CN=.*/"

It means that anyone that has the certs could pretend to be one of your nodes.  I have it on the list of things to dig into at some point.

DandyDeveloper commented:
@justinrush For the time being I've done the wildcard to just bypass as the SSL isn't super important for this phase of what I'm doing. However, it's plausible I was missing an attribute. I'll keep investigating, maybe add some steps in a PR if there's something that could be clearer.

sc7565 commented:
Is there support for latest Open Distro version 1.0.0

DandyDeveloper commented:
@sc7565 God no, but I'm sure that with a bit of effort. This chart can be upgraded to support 1.0.

henrytran026 commented:
> @justinrush For the time being I've done the wildcard to just bypass as the SSL isn't super important for this phase of what I'm doing. However, it's plausible I was missing an attribute. I'll keep investigating, maybe add some steps in a PR if there's something that could be clearer.
> 
> One update: I'm on the homestretch but I'm now facing issues with the client ingress. Giving a 502 because:
> 
> `Someone (null) speaks http plaintext instead of ssl, will close the channel`
> 
> This is with forced SSL over HTTPS. May be a certificates thing my side though, investigating.

I'm also seeing this issue while trying to use the official chart from elastic. Hope it will get resolve soon!

kfox1111 commented:
What ingress controller / annotation?

henrytran026 commented:
> What ingress controller / annotation?

We use nginx controller in our cluster.

kfox1111 commented:
Try nginx.ingress.kubernetes.io/backend-protocol: "HTTPS". There was a different way to do it that worked with older ingress-nginx, but not the newer ones. Bumped into that a little while back.

kfox1111 commented:
Is there a reason not to merge this and then iterate in follow on pr's?

DerekHeldtWerle commented:
Hey everyone, I've updated the chart to support `1.X.X`. I've setup a public gist [here](https://gist.github.com/DerekHeldtWerle/452a31b5fbcbfbaff6dc9328d826a957) describing the major changes and the process I took in order to migrate our `securityconfig` files to comply with the updated version. Due to the various changes made in order to move to the updated version (and to create a more robust and stable chart in general), we approached this more so as starting anew versus an official _upgrade_ due to masters lacking a statefulset + persistent volume. Future looking, we expect future upgrades to be much more seamless due to the aforementioned changes.

gezb commented:
@DerekHeldtWerle  I have just tried your latest commit to this PR as I am just getting started with ODES. I have spotted two issues so far. You seem to have a StatefulSet and Deployment for the master  nodes and the StatefulSet is missing the mandatory "seviceName" entry

rishabh6788 commented:
Hi,

I tried installing es open distro using the helm commands. 
❯ helm package .
❯ helm install opendistro-es-0.0.1.tgz --name opendistro-es
Above commands worked without any issue but the system kind of hanged after that. kubectl commands are timing out and not responding.

Could it be due to my node configuration? I am testing on my mac on minikube node?

sburton84 commented:
It might be good if there was a value that could be set to disable the initContainers which run `sysctl`, in case of running in an environment that doesn't allow privileged containers.

ferozsalam commented:
In addition to the missing `serviceName` entry on the StatefulSet, the presence of both the Deployment and the StatefulSet manifests is causing contention issues on the PVC. My understanding is that only the StatefulSet is required, and dropping the deployment makes the entire cluster work for me.

DandyDeveloper commented:
@ferozsalam Can you provide more details on this? 

With default settings, I haven't had any issues with my PVC creation. That being said, I'm not using static host bound PVCs for my ES cluster, so it's been a while since I've looked at this side of the chart.

scottcressi commented:
This really should be merged.  In general git commits should be small, and while I understand this project/PR is not fully completed this PR is massive and continues to be a moving target.  We should merge so others can resume submitting small piecemeal commits to improve things much faster to reach completion.

DerekHeldtWerle commented:
@DandyDeveloper @scottcressi @alolita, I'm all for getting this merged and moving more towards the model of updates others have mentioned, but we haven't heard from amazon's side from quite some time now. We (and I'm sure others) have been running this without issue for a couple months now, but I'm not sure if there's other things that need to be done in order to get the final approval.

 In the meantime, any and all PR's are welcome against our fork to continue to improve the chart.

scottcressi commented:
> @DandyDeveloper @scottcressi @alolita, I'm all for getting this merged and moving more towards the model of updates others have mentioned, but we haven't heard from amazon's side from quite some time now. We (and I'm sure others) have been running this without issue for a couple months now, but I'm not sure if there's other things that need to be done in order to get the final approval.
> 
> In the meantime, any and all PR's are welcome against our fork to continue to improve the chart.

Does amazon control the merge process?

cq-z commented:
have a error:
![image](https://user-images.githubusercontent.com/3038490/66650789-04a02680-ec64-11e9-8a18-d4aca1b3522a.png)
need delete opendistro-es/templates/elasticsearch/es-master-deploy.yaml 
```
      - name: http.enabled
          value: "true"
```

ferozsalam commented:
@DandyDeveloper yes - the issue was introduced by other (unrelated) changes to support the 1.0.0 image. I notice that @DerekHeldtWerle has just pushed a commit that removes the master deployment, which means that this now works out of the box for me.

DerekHeldtWerle commented:
@scottcressi This repo is indeed controlled by amazon. 

In moving the files from our internal repo to this public one, I had forgotten to remove the master deployment file. This should now be fixed @cq-z as @ferozsalam has pointed out.

cq-z commented:
is ok . thank you!

rishabh6788 commented:
Has anyone tested the above helm chart using custom certs?
If yes, could you please share changes you made to make it work?
I need to know how did you bootstrap the secrets (the certificate content)?

tfrench93 commented:
Just wondering if anyone has any tips for dealing with index management while using this helm chart? (e.g., deleting indices after a certain number of days, etc)

fsschmitt commented:
@tfrench93, I would recommend you to take a look into the [elasticsearch-curator](https://github.com/helm/charts/tree/master/stable/elasticsearch-curator) helm chart, as it will deploy a CronJob (e.g. every day at 2am) to delete certain indexes (regexp) after X amount of days.

sc7565 commented:
> Just wondering if anyone has any tips for dealing with index management while using this helm chart? (e.g., deleting indices after a certain number of days, etc)

Another option is, rebuild the OD with basic license.

tfrench93 commented:
> @tfrench93, I would recommend you to take a look into the [elasticsearch-curator](https://github.com/helm/charts/tree/master/stable/elasticsearch-curator) helm chart, as it will deploy a CronJob (e.g. every day at 2am) to delete certain indexes (pattern) after X amount of days.

@fsschmitt Does curator work well with the open source distributions of ES? I figured it would only pair nicely with the official Elastic distribution!

DerekHeldtWerle commented:
@tfrench93 We have been using the elasticsearch-curator helm chart without issue with opendistro since the onset 👍 

@rishabh6788 We use custom certs and the README has an [example](https://github.com/Viasat/opendistro-elastic-community/blob/feat/helm-chart/open-distro-elasticsearch-kubernetes/helm/README.md#example-secure-elasticsearch-config-with-custom-certs) you can reference as well.

tfrench93 commented:
Awesome! Thanks for the info @DerekHeldtWerle :)

DandyDeveloper commented:
> Why hasn't this merged yet? Its been open since May 2.

@kfox1111 Read up. 

https://github.com/opendistro-for-elasticsearch/community/pull/56#issuecomment-540105926

waynetaylor commented:
Observation here RE the helm chart. I am still new to K8's so learning but i was getting. Master and data nodes pending. After debugging I saw following:

kubectl describe pods opendistro-es-1-data-0

Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  43s   default-scheduler  pod has unbound immediate PersistentVolumeClaims (repeated 3 times)
I apply a storage class to satisfy but this seems missing in the helm package

alolita commented:
Hi @DerekHeldtWerle @kalvinnchau - We're in the process of reviewing final commits and will ask on thread for clarifications. Thanks again for your patience!

alolita commented:
Hi @DerekHeldtWerle, @kalvinnchau,

Can you please make sure all files (including .yaml files) are licensed with a Apache-2 license not MIT-0. Thanks!

alolita commented:
@DerekHeldtWerle, @kalvinnchau,

Please change description (e.g. in Chart.yaml) from description: 'Opendistro Elasticsearch' to 'Open Distro for Elasticsearch'?

DandyDeveloper commented:
@DerekHeldtWerle  @kalvinnchau 
https://github.com/Viasat/opendistro-elastic-community/pull/10

I've just thrown in a relatively small, but important change in my org that I think others would benefit from. Let me know your thoughts in the PR.

DerekHeldtWerle commented:
Hey @alolita, just to double check, is the below in line with what you are looking for in terms of licensing for each file?:

```
# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
```

@DandyDeveloper I'll look that over today.

alolita commented:
Hi @DerekHeldtWerle can you please use the header in https://github.com/opendistro-for-elasticsearch/perftop/blob/master/bin/global.js as an example. :-)

DerekHeldtWerle commented:
Hey @alolita, the licenses have been added to all files. If there's anything that needs changing let me know 🙂

DerekHeldtWerle commented:
Hey @alolita, is there anything else you need from our end? I'd love to merge it but don't have write access 😄

rishabh6788 commented:
LGTM. Merging this PR.
Please keep providing your feedbacks and keep improving this feature.

Thank you!
